src/ring_theory/polynomial/content.lean

```diff
@@ -0,0 +1,307 @@
+end
+
+lemma C_content_dvd (p : polynomial R) : C p.content ∣ p :=
+dvd_content_iff_C_dvd.1 (dvd_refl _)
+
+/-- A polynomial over a GCD domain is primitive when the `gcd` of its coefficients is 1 -/
+def is_primitive (p : polynomial R) : Prop := p.content = 1
+
+@[simp]
+lemma is_primitive_one : is_primitive (1 : polynomial R) :=
+by rw [is_primitive, ← C_1, content_C, normalize_one]
+
+lemma is_primitive.ne_zero {p : polynomial R} (hp : p.is_primitive) : p ≠ 0 :=
+begin
+  rintro rfl,
+  rw [is_primitive, content_zero] at hp,
+  apply zero_ne_one hp,
+end
+
+lemma is_primitive.content_eq_one {p : polynomial R} (hp : p.is_primitive) : p.content = 1 := hp
+
+lemma is_primitive_iff_is_unit_of_C_dvd {p : polynomial R} :
+  p.is_primitive ↔ ∀ (r : R), C r ∣ p → is_unit r :=
+begin
+  rw [is_primitive],
+  split,
+  { intros h r hdvd,
+    rw [← dvd_content_iff_C_dvd, h] at hdvd,
+    apply is_unit_of_dvd_one _ hdvd },
+  { intro h,
+    rw [← normalize_content, normalize_eq_one],
+    apply h _ (C_content_dvd _) }
+end
+
+lemma eq_C_mul_primitive (p : polynomial R) :
+  ∃ (r : R) (q : polynomial R), p = C r * q ∧ q.is_primitive ∧ p.nat_degree = q.nat_degree :=
+begin
+  by_cases h0 : p = 0,
+  { use [0, 1],
+    simp [h0] },
+  rcases C_content_dvd p with ⟨q, h⟩,
+  use [p.content, q],
+  refine ⟨h, _, _⟩,
+  { have h1 := content_C_mul p.content q,
+    rw [← h, normalize_content, ← mul_one p.content, mul_assoc, one_mul] at h1,
+    apply mul_left_cancel' _ h1.symm,
+    rwa ← content_eq_zero_iff at h0 },
+  { rw [h, mul_eq_zero] at h0,
+    classical,
+    rcases (decidable.not_or_iff_and_not _ _).1 h0,
+    rw [h, nat_degree_mul left right, nat_degree_C, zero_add] }
+end
+
+lemma gcd_content_eq_of_dvd_sub {a : R} {p q : polynomial R} (h : C a ∣ p - q) :
+  gcd_monoid.gcd a p.content = gcd_monoid.gcd a q.content :=
+begin
+  rw content_eq_gcd_range_of_lt p (max p.nat_degree q.nat_degree).succ
+    (lt_of_le_of_lt (le_max_left _ _) (nat.lt_succ_self _)),
+  rw content_eq_gcd_range_of_lt q (max p.nat_degree q.nat_degree).succ
+    (lt_of_le_of_lt (le_max_right _ _) (nat.lt_succ_self _)),
+  apply finset.gcd_eq_of_dvd_sub,
+  intros x hx,
+  cases h with w hw,
+  use w.coeff x,
+  rw [← coeff_sub, hw, coeff_C_mul]
+end
+
+lemma content_mul_aux {p q : polynomial R} :
+  gcd_monoid.gcd (p * q).erase_lead.content p.leading_coeff =
+  gcd_monoid.gcd (p.erase_lead * q).content p.leading_coeff :=
+begin
+  rw [gcd_comm (content _) _, gcd_comm (content _) _],
+  apply gcd_content_eq_of_dvd_sub,
+  rw [← self_sub_C_mul_X_pow, ← self_sub_C_mul_X_pow, sub_mul, sub_sub, add_comm, sub_add,
+    sub_sub_cancel, leading_coeff_mul, ring_hom.map_mul, mul_assoc, mul_assoc],
+  apply dvd_sub (dvd.intro _ rfl) (dvd.intro _ rfl),
+end
+
+@[simp]
+theorem content_mul {p q : polynomial R} : (p * q).content = p.content * q.content :=
+begin
+  have h : ∀ (n : ℕ) (p q : polynomial R), (p * q).nat_degree ≤ n
+    → (p * q).content = p.content * q.content,
+  { clear p q,
```

Comment:
This proof is a bit of a monster...
One little tweak would be to replace the `have h` above by a `suffices h`,
and then prove the main goal in 1 line.
After that, you only need to prove `h`, and you don't need to indent and use `{}` for that.
Saves you one level of indentation.